elements/lisk-tree/src/merkle_tree/merkle_tree.ts

```diff
@@ -23,42 +23,57 @@ import {
 	EMPTY_HASH,
 	LEAF_PREFIX,
 	BRANCH_PREFIX,
+	PREFIX_HASH_TO_VALUE,
+	PREFIX_LOC_TO_HASH,
 } from './constants';
-import { NodeData, NodeInfo, NodeType, NodeSide, Proof } from './types';
+import { InMemoryDB } from './inmemory_db';
+import { PrefixStore } from './prefix_store';
+import { NodeData, NodeInfo, NodeType, NodeSide, Proof, Database } from './types';
 import { generateHash, getBinaryString, isLeaf, getPairLocation } from './utils';
 
 export class MerkleTree {
 	private _root: Buffer;
-	private _width = 0;
+	private _size = 0;
 	private readonly _preHashedLeaf: boolean;
+	private readonly _db: Database;
 
-	// Object holds data in format { [hash]: value }
-	private _hashToValueMap: { [key: string]: Buffer | undefined } = {};
-	private _locationToHashMap: { [key: string]: Buffer | undefined } = {};
+	private readonly _hashToValueMap: PrefixStore;
+	private readonly _locationToHashMap: PrefixStore;
 
-	public constructor(initValues: Buffer[] = [], options?: { preHashedLeaf: boolean }) {
+	public constructor(options?: { preHashedLeaf?: boolean; db?: Database }) {
+		this._db = options?.db ?? new InMemoryDB();
+		this._hashToValueMap = new PrefixStore(this._db, PREFIX_HASH_TO_VALUE);
+		this._locationToHashMap = new PrefixStore(this._db, PREFIX_LOC_TO_HASH);
+		this._preHashedLeaf = options?.preHashedLeaf ?? false;
+
+		this._root = EMPTY_HASH;
+	}
+
+	public async init(initValues: Buffer[]): Promise<void> {
```

Comment:
why do we need seperate `init` here? why not just in constructor?

Reply:
it's needed because now get/set is `Promise` and construtor cannot be async

Reply:
I guess because constructors can't be async. The constructor must return an object of a particular type only, means `MerkleTree` constructor must return an object of type `MerkleTree`, making it async will turn type to `Promise<MerkleTree>`, so we can't make constructors async.